src/lib/operators/table_scan.hpp

```diff
@@ -25,7 +28,165 @@ class TableScan : public AbstractOperator {
+                                                        std::shared_ptr<const Table>& real_input_table) const;
+
+  template <typename T>
+  bool _evaluate_scan_predicate(const T& search_value, const T& value) const {
+    switch (_scan_type) {
+      case ScanType::OpEquals: {
+        return value == search_value;
+      }
+      case ScanType::OpNotEquals: {
+        return value != search_value;
+      }
+      case ScanType::OpLessThan: {
+        return value < search_value;
+      }
+      case ScanType::OpLessThanEquals: {
+        return value <= search_value;
+      }
+      case ScanType::OpGreaterThan: {
+        return value > search_value;
+      }
+      case ScanType::OpGreaterThanEquals: {
+        return value >= search_value;
+      }
+      default: {
+        throw std::logic_error("Scan type not supported");
+      }
+    }
+  }
+
+  bool _evaluate_scan_predicate_for_dictionary_segment(const ValueID value_id, const ValueID lower_bound,
+                                                       bool value_exists) const {
+    switch (_scan_type) {
+      case ScanType::OpEquals: {
+        if (!value_exists) {
+          return false;
+        } else {
+          return value_id == lower_bound;
+        }
+      }
+      case ScanType::OpNotEquals: {
+        if (!value_exists) {
+          return true;
+        }
+        return value_id != lower_bound;
+      }
+      case ScanType::OpLessThan: {
+        return value_id < lower_bound;
+      }
+      case ScanType::OpLessThanEquals: {
+        if (value_exists) {
+          return value_id <= lower_bound;
+        } else {
+          return value_id < lower_bound;
+        }
+      }
+      case ScanType::OpGreaterThan: {
+        if (value_exists) {
+          return value_id > lower_bound;
+        } else {
+          return value_id >= lower_bound;
+        }
+      }
+      case ScanType::OpGreaterThanEquals: {
+        return value_id >= lower_bound;
+      }
+      default: {
+        throw std::logic_error("Scan type not supported");
+      }
+    }
+  }
+
+  template <typename T>
+  void _collect_matched_rows_for_reference_segment(const std::shared_ptr<ReferenceSegment>& segment,
+                                                   const std::shared_ptr<PosList>& matched_row_ids) const {
+    for (auto const& row_id : *(segment->pos_list())) {
+      const ChunkID chunk_id = row_id.chunk_id;
+      const ChunkOffset real_chunk_offset = row_id.chunk_offset;
+
+      const ColumnID column_id = segment->referenced_column_id();
+      auto table = segment->referenced_table();
+
+      const Chunk& chunk = table->get_chunk(chunk_id);
+      const auto actual_segment = chunk.get_segment(column_id);
+
+      const std::string column_type = table->column_type(_column_id);
+
+      resolve_data_type(column_type, [&](auto type) {
+        using Type = typename decltype(type)::type;
+        const Type search_value = type_cast<Type>(_search_value);
```

Comment:
Great that you are using the ReferenceSegment's underlying segment but there are ways to avoid using `resolve_data_type` and `type_cast` for every single row.